### ISSUE TITLE
Capacitor Support

### ISSUE THREAD

**Issue by sandstrom** [Pushwoosh/pushwoosh-phonegap-plugin#290]
Curious to hear if you have verified if your plugin works with Capacitor (an open-source project based on, and similar to, Cordova)?

https://github.com/ionic-team/capacitor

**Comment by wfhm** [comment]
Hi,

We did not check if our plugin works with Capacitor yet. Have you already tried using the plugin with this framework on your own?

**Comment by chintan13** [comment]
is this worked for anyone in capacitor?

**Comment by sandstrom** [comment]
@chintan13 I haven't tried yet, sorry 😄

**Comment by sandstrom** [comment]
@chintan13 I've reopened, but I think it would help if you'd add more details for the Pushwoosh team to troubleshoot.

**Comment by chintan13** [comment]
For Android its working fine

For ios if npx cap sync not working then we just need to comment out or add target="config.xml"
`<config-file>
    <access origin="*.pushwoosh.com" />
</config-file> `

and then try to build from xcode and it gives below error

ld: library not found for -lPushwoosh
clang: error: linker command failed with exit code 1 (use -v to see invocation)

**Comment by wfhm** [comment]
@chintan13 @sandstrom Capacitor framework is now supported starting with 7.12.0 release.

**Comment by sandstrom** [comment]
*(for anyone who's curious, here is the PR that added support: https://github.com/Pushwoosh/pushwoosh-phonegap-plugin/pull/306)*

**Comment by ipehimanshu** [comment]
hi @wfhm 

can you please share few steps how to make it working ?

I mean what we need to follow steps,

like what file we need to put into android folder,
what we need to edit for xcode/ios platform ,

Thanks a lot if you share some more info specially for capacitor 

Thanks a lot

**Comment by wfhm** [comment]
@ipehimanshu 
I think I will drop a working sample on Github later this week. Meanwhile, may I ask you what particular issues did you have when trying to integrate the plugin?

**Comment by ipehimanshu** [comment]
Hi

Actually i am not able to get push in android when app already open
and for ios we try to run app with xcode (real device) but it never
received push, so i am not sure but we failed to make it working, same code
working with cordova,
so we really appreciate if you make one sample with capacitor for both
platform

Thanks a lot

**Comment by wfhm** [comment]
@ipehimanshu If I understood you correctly, you just migrated your Cordova app to Capacitor, is that so? If yes, then it doesn't really make sense to create a sample, since the integration process is different.

> Actually i am not able to get push in android when app already open
This is actually the default behavior of our plugin. Could it be that you have previously installed the plugin in your Cordova app with the following variable?

`--variable ANDROID_FOREGROUND_PUSH="true"`

If so, you can do the same with Capacitor by adding this flag to capacitor.config.json:
```
"cordova": {
    "preferences": {
        "ANDROID_FOREGROUND_PUSH"=true
    }
  }
```
Alternatively, you can set the corresponding meta-data to AndroidManifest.xml directly:
```
<meta-data android:name="com.pushwoosh.foreground_push" android:value="true" />
```

> and for ios we try to run app with xcode (real device) but it never
received push

Could you please share us with the device console logs so we could check it on our side? Just launch the app from Xcode, subscribe for pushes, send a push and collect the log (even if no push is received)?

**Comment by ipehimanshu** [comment]
Hello

We already try

"cordova": {
    "preferences": {
        "ANDROID_FOREGROUND_PUSH"=true
    }
  }

but it never update AndroidManifest.xml because capacitor use
capacitor-cordova-android-plugins and it also has AndroidManifest.xaml
and i see it value never update manually
[image: image.png]

If i edit manually that value as false to true, then it works but once
i fire command like npx cap sync that xml file again update and it set
value as false again.

So that is reason we are stuck

and if i try to add into android folder AndroidManifest.xml then it
give error like merge resources meta data

Hope it will be more clear for you

FYI current we are using capacitor latest version (4)

Thank you

**Comment by wfhm** [comment]
> [image: image.png]
Looks like the image got lost:)
Apparently, Capacitor does not support Cordova install variables at all:
> Capacitor does not support Cordova install variables, auto configuration, or hooks, due to our philosophy of letting you control your native project source code (meaning things like hooks are unnecessary). If your plugin requires variables or settings to be set, you'll need to apply those configuration settings manually by mapping between the plugin's plugin.xml and required settings on iOS and Android.

The plugin currently does not provide any other way to change this value, so you will have to manually do it in the plugin itself. You can find the plugin source code in your node_modules directory and modify the **[PushwooshNotificationServiceExtension](https://github.com/Pushwoosh/pushwoosh-phonegap-plugin/blob/master/src/android/src/com/pushwoosh/plugin/pushnotifications/PushwooshNotificationServiceExtension.java)** 
class by replacing the code below

```
@Override
	protected boolean onMessageReceived(final PushMessage pushMessage) {
		PushNotifications.messageReceived(pushMessage.toJson().toString());
		return (!showForegroundPush && isAppOnForeground()) || super.onMessageReceived(pushMessage);
	}
```
with this:
```
@Override
	protected boolean onMessageReceived(final PushMessage pushMessage) {
		PushNotifications.messageReceived(pushMessage.toJson().toString());
		return super.onMessageReceived(pushMessage);
	}
```
We'll update the plugin to provide an easier option to control foreground notifications behavior, meanwhile, you can try the workaround above. If you give it a shot, please let us know about the results.

**Comment by ipehimanshu** [comment]
ok so 
@wfhm possible to create issue with capacitor team 

here is image screenshot which i want to share
https://prnt.sc/4NikmzKS15lZ

Thank you

**Comment by wfhm** [comment]
@ipehimanshu I don't think we need Capacitor team at this moment - the plugin's design was not adapted to Capacitor according to their guidelines, so it should be done on our side anyway.

**Comment by ipehimanshu** [comment]
ok got it

**Comment by ipehimanshu** [comment]
@wfhm 
i confirm like after edit that file it comes foreground push for android also

Thank you

**Comment by ajisakson** [comment]
I'd like to make note here that the Cordova Pushwoosh plugin will not automatically add the required intents to your AndroidManifest.xml, so you'll need to add them yourself. For me this meant putting the following somewhere within the <application> tags:

`<meta-data android:name="com.pushwoosh.log_level" android:value="DEBUG"/>
    <meta-data android:name="com.pushwoosh.foreground_push" android:value="false"/>
    <meta-data android:name="com.pushwoosh.notification_service_extension" android:value="com.pushwoosh.plugin.pushnotifications.PushwooshNotificationServiceExtension"/>
    <meta-data android:name="com.pushwoosh.internal.plugin_provider" android:value="com.pushwoosh.plugin.internal.PhonegapPluginProvider"/>
    <service android:exported="false" android:name="com.pushwoosh.plugin.pushnotifications.CustomFirebaseMessagingService">
      <intent-filter>
        <action android:name="com.google.firebase.MESSAGING_EVENT"/>
      </intent-filter>
    </service>`
    
This will make sure that the documented Pushwoosh event listeners work and that you can receive the notification when the app is killed or in the background on Android.

**Comment by enginseer-dev** [comment]
@ajisakson, thank you for your comment! Please let me check it on my own and update docs or the plugin if required.

**Comment by enginseer-dev** [comment]
Hello, @ajisakson 

Thank you for your patience.

We've integrated the Pushwoosh Cordova plugin in a sample Capacitor project.
The necessary intents are automatically added to the AndroidManifest.xml during the build process, as verified in the decompiled APK:
![image](https://github.com/Pushwoosh/pushwoosh-phonegap-plugin/assets/139750027/3b0bf392-d571-4ddf-ad64-9c964ff020a0)

This ensures that push notifications are correctly received even when the app is in the background or killed.

Please let us know if there are any further questions or concerns.

**Comment by ajisakson** [comment]
@enginseer-dev
Can you confirm if this is how it has and should have been working previously? Or is this a recent addition to the plugin? 

Either way, glad to see it working as such and I will make sure on our next Capacitor project with Pushwoosh that this is the case and will report back if we still have issue.

Thanks!

**Comment by wfhm** [comment]
@ajisakson this is the expected behavior, and it's been there for quite a long time. The metadata and other components of our SDK are merged into the final application manifest during the build process, and it is done by Gradle, not by Cordova/Capacitor tools, so it's quite strange that you had to do it manually. Should you have the same issue in your new projects, please let us know and we'll look into this again, but for now everything seems to be working as expected.

**Comment by ipehimanshu** [comment]
Hello Team,

Is it fixed issue which i need to manually change

PushwooshNotificationServiceExtension.java  (because capacitor not allow to
change false to true as we need to always receive push for all cases)
//return (!showForegroundPush && isAppOnForeground()) ||
super.onMessageReceived(pushMessage);
    return super.onMessageReceived(pushMessage);

**Comment by wfhm** [comment]
@ipehimanshu you can set the `ANDROID_FOREGROUND_PUSH` plugin variable to **true** or **false** in your `capacitor.config.json`.

**Comment by ipehimanshu** [comment]
ok so will it automatic set true to androidmanifest.xml 
because so far i remember you says to me like edit some core files

@wfhm please show one example which is working like that with capacitor ?

**Comment by enginseer-dev** [comment]
@ipehimanshu I have tested and the flags aren't working yet.
The following manual edit is still needed for the foreground push notifications to work:

> You can find the plugin source code in your node_modules directory and modify the [PushwooshNotificationServiceExtension](https://github.com/Pushwoosh/pushwoosh-phonegap-plugin/blob/master/src/android/src/com/pushwoosh/plugin/pushnotifications/PushwooshNotificationServiceExtension.java)
> class by replacing the code below
> 
```
@Override
	protected boolean onMessageReceived(final PushMessage pushMessage) {
		PushNotifications.messageReceived(pushMessage.toJson().toString());
		return (!showForegroundPush && isAppOnForeground()) || super.onMessageReceived(pushMessage);
	}
```
> with this:
> 
```
@Override
	protected boolean onMessageReceived(final PushMessage pushMessage) {
		PushNotifications.messageReceived(pushMessage.toJson().toString());
		return super.onMessageReceived(pushMessage);
	}
```

We will look into this issue and will fix it in the next release.

**Comment by ipehimanshu** [comment]
ok please


Thank you

**Comment by enginseer-dev** [comment]
Hello, @ipehimanshu 

We've released a new version of the [Cordova plugin 8.3.23](https://github.com/Pushwoosh/pushwoosh-phonegap-plugin/releases/tag/8.3.23). 
The method allows you to configure the display of push notifications while the app is in the foreground by setting it to either false or true.

Example:

```
function initPushwoosh() {
    var pushwoosh = cordova.require("pushwoosh-cordova-plugin.PushNotification");

    // Should be called before pushwoosh.onDeviceReady
    document.addEventListener('push-notification', function(event) {
        var notification = event.notification;
        // Handle push open here
        console.log('Received push notification: ', notification);
    });
 
    // Initialize Pushwoosh. This will trigger all pending push notifications on start.
    pushwoosh.onDeviceReady({
        appid: "APP_CODE",
        projectid: "SENDER_ID",
        serviceName: "MPNS_SERVICE_NAME" // Omit this if not using MPNS
    });

    // Registering for push notifications
    pushwoosh.registerDevice(
        function(status) {
            var pushToken = status.pushToken;
            // Handle successful registration here
            console.log('Push token received: ', pushToken);
        },
        function(status) {
            // Handle registration error here
            console.error('Push registration failed: ', status);
        }
    );

    pushwoosh.setShowPushnotificationAlert(true);
}
```
